### ISSUE TITLE
NaN, +Inf, -Inf, Inf

### ISSUE THREAD

**Issue by gvanuxem** [PDLPorters/pdl#309]
Hello,

What about adding routines that return these values in PDL::Core?

I know they were available via PDL::LinearAlgebra but they are more general.

Regards,

G. Vanuxem

**Comment by mohawk2** [comment]
As discussed, it will be so! Although to be clear, do you want different things for `inf` and `posinf`? Are they synonyms?

**Comment by zmughal** [comment]
Just of note, you can get those values via the constructor:

```perl
use PDL;
my $p = pdl('nan inf -inf');
use DDP; p $p;
# PDL {
#     Data     : [NaN Inf -Inf]
#     Type     : double
#     Shape    : [3]
#     Nelem    : 3
#     Min      : NaN
#     Max      : NaN
#     Badflag  : No
#     Has Bads : No
# }

```

**Comment by gvanuxem** [comment]
No, it's just a matter of sign in a computer I think.

**Comment by zmughal** [comment]
Just to compare:

Octave: <https://octave.sourceforge.io/octave/function/Inf.html>, <https://octave.sourceforge.io/octave/function/NaN.html>
```octave
octave:1> [ inf, -inf, Inf, -Inf, nan, NaN ]
ans =

   Inf  -Inf   Inf  -Inf   NaN   NaN
```

NumPy: <https://numpy.org/devdocs/reference/constants.html>
```python
In [1]: import numpy as np

In [2]: [ np.inf, np.PINF, np.Inf, np.infty, np.NINF, np.nan, np.NaN ]
Out[2]: [inf, inf, inf, inf, -inf, nan, nan]
```

**Comment by mohawk2** [comment]
I think making `inf` and `nan` take dimensions and act like `ones` would be very useful. My feeling is that negative infinity is actually more idiomatically achieved by negating the return value of our hypothetical `inf`. What do you folks think?

**Comment by zmughal** [comment]
Yes, I agree about taking dimensions and about negating the return value.

I do have a question about how this would work with Perl's built-in support for Inf and NaN. Would it be necessary to use the namespace (i.e., not exported by default).


```perl
say 0 + NaN;
# NaN
say 0 + Inf;
# Inf
say 0 - Inf;
# -Inf

### But with a sub

sub Inf { 42 }; say 0 + Inf
# 42
```

**Comment by mohawk2** [comment]
My thinking was lower-case, and exported by default like any other function.

**Comment by zmughal** [comment]
Sounds good to me on all points:

`inf()`, `nan()`

- take dimensions
- lower-case
- exported by default
- use `- inf(x,y, ...)` to get ndarray of `-Inf`

**Comment by mohawk2** [comment]
Currently `zeroes` blows up without arguments, that could usefully be fixed at the same time!

**Comment by zmughal** [comment]
Sounds good. Yay for consistency :-)

**Comment by gvanuxem** [comment]
Hello,

When used at the C level directly:

pdl> use PDL::LinearAlgebra
pdl> p nan()
NaN
pdl> p posinf()
Inf
pdl> p neginf()
-Inf
pdl>

Just my two cents.

Greg

**Comment by mohawk2** [comment]
Yes, I'm looking at how to make it return a scalar with no args. Since `zeroes` and `ones` blow up with no args, there seems to be room for a new consistent behaviour.

**Comment by mohawk2** [comment]
Implemented in 938756271a585ff5c00a461ac4930414cd9f6db8. Review comments welcome!

**Comment by mohawk2** [comment]
Released as 2.042. @gvanuxem I don't see any need to update PDL::LinearAlgebra as it works fine.

**Comment by gvanuxem** [comment]
Yes, you're right. It used to not work. For sure.
